### PR TITLE
AnimateCSS

### PR DESCRIPTION
**Pull request by bugsounet**
Hi,

This is my testing code for AnimateCSS for `show()`, `hide()`, `updateDom()`

Naturally, we have to do better !

I voluntarily modify `newsfeed` and `compliments` in order to test

Note: I will correct checks later... it's a test...

### CODE REVIEW

**Comment by bugsounet** [comment]
ok... I will solve localy :/

**Comment by bugsounet** [comment]
So...

I purpose to you to make a test with `newsfeed` module
just change lines 304 and 308 with : `this.updateDom(this.config.animationSpeed, 3, 14);`

syntax: `updateDom(speed, animateOut, animateIn)`
animateOut, animateIn number value are available [there](https://github.com/bugsounet/MagicMirror/blob/develop/js/animateCSS.js) and it's the result of the animated names from the [Animate.style website](https://animate.style/)

**Comment by khassel** [comment]
> Result: The news will come from the left, wait in the middle and come out the right

tested this with the same result

**Comment by bugsounet** [comment]
@rejas says `Start with a very simple animation so we can discuss the actual implementation?`

I know, I have to do better implementation
It's just a testing code actually.

Should I do better? (Of course)
Are you really interested in this optional feature?

**Comment by rejas** [comment]
> @rejas says `Start with a very simple animation so we can discuss the actual implementation?`
> 
> I know, I have to do better implementation It's just a testing code actually.
> 
> Should I do better? (Of course) Are you really interested in this optional feature?

Sorry for not being more active on your PR, I really like the feature but with a newborn I cant concentrate right now much on these PRs (contrary to small stuff like dependency updates that dont take much brain power)

Expect more feedback next week, then I might have more time to think about it.

**Comment by bugsounet** [comment]
@rejas: congratulations to you ! This is happiness !

Take your time, a newborn is more important than this PR (I know the time and energy it takes!)

**Comment by bugsounet** [comment]
I will try to do something cleaner because a test code is not made to be in production

**Comment by bugsounet** [comment]
I realize that it's not that easy to do something cleaner

**Comment by rejas** [comment]
> just change lines 304 and 308 with : `this.updateDom(this.config.animationSpeed, 3, 14);`

Would it be possible to have something more "verbose" when calling the updateDom? Something more like `this.updateDom(this.config.animationSpeed, Animate.Bounce, Animate.FadeOut);`
which would make it more readable for developers?

Also what do you think about having the module animation configurable (just like position):
```
		{
			module: "clock",
			position: "top_left",
                        animateIn: "bounce",
                        animateOut: "fade_out",
		},
```

?

As for the cleanliness of the code: I dont mind testing code in there, as long as the API to the modules is well-thought of. The actual implementation can be changed easily later, the API stays for a long time :-)

**Comment by bugsounet** [comment]
Oh yes!!! very good idea.
I will try to do something about it

You prefer `Animate.Bounce` , `Animate.FadeOut`, `Animate.xxx`
I will create an object to manage this then

On the other hand, what bothers me, if we put all the animations in the same object that would mean that we can use the "animateOut" in the "animateIn" and vice versa
That's why I separated the animations

I think you would like that if the name of the requested animation does not exist, we fall back on the default MM² animation

Warning:
  * don't forget that animations add a class at the module position (it's add `animate__animated <animation name>` classes)
  * this necessarily breaks the test suite
  * You will have to review your test suite :/

**Comment by bugsounet** [comment]
(sometime e2e scares me...)

**Comment by bugsounet** [comment]
Right !

So i have added like you have says:

```
		{
			module: "compliments",
			position: "lower_third",
			animateIn: "bounceIn",
			animateOut: "bounceOut"
		},
```

`this.updateDom(this.config.animationSpeed, "bounceOut" , "bounceIn");` is available too

`this.hide(<speed>, { animate: <animateName> })`
and naturally the same with `show()`

Note: if a user have defined `animateIn` and/or `animateOut` in the module definition : MM² with use it in priority and will not apply `updateDom`, `hide`, `show` rules coded by the developer
I think this allows the end user to choose what he wants to have as animation with his own configuration

**Comment by bugsounet** [comment]
Just because humm... I'm crazy !

I just add possibility to use animateIn on start of MM²

@rejas : what do you think about that ?

**Comment by bugsounet** [comment]
@rejas : Veeck,
this must be good for me

If you want to see again ;)

File: js/animateCSS.js [hunk 1]
```diff
@@ -0,0 +1,165 @@
+/* MagicMirror²
+ * AnimateCSS System from https://animate.style/
+ * by @bugsounet
+ * for Michael Teeuw https://michaelteeuw.nl
+ * MIT Licensed.
+ */
+
+/* enumeration of animations in Array **/
+const _AnimateCSSIn = [
```
Comment (rejas) [inline js/animateCSS.js:9]: that _ in front of the name is a Pattern we dont use anywhere else. what should it signal?

File: js/animateCSS.js [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+	"zoomInRight",
+	"zoomInUp",
+	// Sliding entrances
+	"slideInDown",
+	"slideInLeft",
+	"slideInRight",
+	"slideInUp"
+];
+
+const _AnimateCSSOut = [
```
Comment (rejas) [inline js/animateCSS.js:77]: same here, why the _ ?

File: js/main.js [hunk 3]
```diff
@@ -22,6 +22,10 @@ const MM = (function () {
 				return;
 			}
 
+			var haveAnimateIn = null;
```
Comment (rejas) [inline js/main.js:25]: Please use let/const instead of var in your code

File: js/main.js [hunk 4]
```diff
@@ -515,8 +590,10 @@ const MM = (function () {
 		 * Update the dom for a specific module.
 		 * @param {Module} module The module that needs an update.
 		 * @param {number} [speed] The number of microseconds for the animation.
+		 * @param {string} [animateOut] AnimateCss animation name before hidden
+		 * @param {string} [animateIn] AnimateCss animation name on show
 		 */
-		updateDom: function (module, speed) {
+		updateDom: function (module, speed, animateOut, animateIn) {
```
Comment (rejas) [inline js/main.js:596]: in a perfect world we would refactor the sighanture to 
(module, {options: {
 speed,
 animate: {
  out,
  in
}) and have a fallback to the old signature. what do you think?

File: js/module.js [hunk 5]
```diff
@@ -205,6 +205,9 @@ const Module = Class.extend({
 		this.name = data.name;
 		this.identifier = data.identifier;
 		this.hidden = false;
+		// is it really useful? (not sure)
```
Comment (rejas) [inline js/module.js:208]: what should it be useful for?

**Review by rejas — CHANGES_REQUESTED** [review]

**Comment by rejas** [comment]
sorry for the long turnaroundtime, lots of stuff to do and the cold is already creeping in...

**Comment by bugsounet** [comment]
no problem, I'm just doing my best for the next release :)

Note: when #3179 is fixed, can you retry test suite?
It's works in local but...

Thx Veeck

**Comment by KristjanESPERANTO** [comment]
With #3181 the test shouldn't fail anymore.

I'm looking forward to this feature. Thanks @bugsounet :slightly_smiling_face:

**Comment by rejas** [comment]
Merged develop myseld, lets see the test results  :-)

**Comment by bugsounet** [comment]
Happy crazy test! 😁

File: js/main.js [hunk 6]
```diff
@@ -101,11 +110,30 @@ const MM = (function () {
...
 	 * @returns {Promise} Resolved when the dom is fully updated.
 	 */
-	const updateDom = function (module, speed) {
+	const updateDom = function (module, updateOptions, createAnimatedDom = false) {
 		return new Promise(function (resolve) {
+			let speed = updateOptions;
+			let animateOut = null;
+			let animateIn = null;
+			if (typeof updateOptions === "object") {
+				if (typeof updateOptions.options === "object" && updateOptions.options.speed !== undefined) {
```
Comment (rejas) [inline js/main.js:123]: is the first typeof check in this line necessary? its already done the line before...

**Review by rejas — CHANGES_REQUESTED** [review]

File: js/main.js (hunk 4 above)
Comment (rejas) [reply js/main.js]: thx for the refactor

**Comment by rejas** [comment]
Ah, screw it, will merge it, we can cleanup further later if necessary (and I dont want to have @bugsounet waiting longer). This is too nice to not have, thx a lot for your work!

**Comment by bugsounet** [comment]
I will start to write MagicMirror Doc

Note: It’s a shame there isn’t a develop branch on the doc
To see the result of the formatting once written (and correct if necessary)

**Comment by bugsounet** [comment]
we will also have to see with @MichMich
it would be perfect if we had a "point of view" develop of this branch

something like "https://dev-docs.magicmirror.builders/" to see develop branch result
just for see if formating is correct before merging it

**Comment by MichMich** [comment]
I've added a `develop` branch to the docs repo which is synced with https://develop.docs.magicmirror.builders. Let me know if there is anything else I can do.

**Comment by bugsounet** [comment]
Thx @MichMich,
It's perfect for the moment !

**Comment by bugsounet** [comment]
@rejas:

I see this bugs:

 * AnimateCSS merge hide() and show() animated css class when we do multiple call

--> result it will stay en hide state

I think event listener (is animateCSS file) is not a proper solution
I will correct soon as possible

**Pull request merged**